packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,186 @@
+// @ts-check
+
+// no-lonely-if is a stupid rule that really should be disabled globally
+/* eslint-disable no-lonely-if */
+
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        Promise.resolve().then(() => {
+          if (wasFulfilled) {
+            if (watcher.onFulfilled) {
+              watcher.onFulfilled(value, ...args);
+            }
+          } else {
+            if (watcher.onRejected) {
+              watcher.onRejected(value, ...args);
+            } else {
+              throw value;
```

Comment:
oh and let's add a comment here, something like `// allow unhandled rejection handler to fire`